README.md
# hf_point_data

A package for doing great things!

## Installation

```bash
$ pip install hf_point_data
```

## Usage

- TODO

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`hf_point_data` was created by Amy C. Defnet. It is licensed under the terms of the MIT license.

## Credits

`hf_point_data` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (08/09/2023)

- First release of `hf_point_data`!

tests/test_hf_point_data.py
import sys
import os
import pytest
import sqlite3
import pandas as pd
import numpy as np

sys.path.append(
    os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../"))
)

from hf_point_data import hf_point_data, utils
from hydrodata.mock_classes.mock_hydrodata import (create_mock_hydrodata, create_mock_observations_database,
                                                   cleanup_mock_hydrodata)

HYDRODATA = 'test_data/hydrodata'


def test_check_inputs_failure1():
    '''Parameter all_attributes cannot be True if return_metadata is False'''
    with pytest.raises(Exception):
        utils.check_inputs(data_source='usgs_nwis', variable='streamflow', temporal_resolution='daily',
                           aggregation='average', return_metadata=False, all_attributes=True)


def test_check_inputs_failure2():
    '''Parameter provided for variable not in supported list (typo).'''
    with pytest.raises(Exception):
        utils.check_inputs(data_source='usgs_nwis', variable='steamflow',
                           temporal_resolution='daily', aggregation='average')


def test_check_inputs_failure3():
    '''Parameter provided for temporal_resolution not in supported list.'''
    with pytest.raises(Exception):
        utils.check_inputs(data_source='usgs_nwis', variable='streamflow',
                           temporal_resolution='monthly', aggregation='average')


def test_get_var_id():
    create_mock_hydrodata(HYDRODATA)
    create_mock_observations_database(HYDRODATA)
    conn = sqlite3.connect(f'{HYDRODATA}/national_obs/point_obs.sqlite')

    # Build SQL connection to mock HydroData database
    assert utils.get_var_id(conn, data_source='usgs_nwis', variable='streamflow',
                            temporal_resolution='hourly', aggregation='average') == 1
    assert utils.get_var_id(conn, data_source='usgs_nwis', variable='streamflow',
                            temporal_resolution='daily', aggregation='average') == 2
    assert utils.get_var_id(conn, data_source='usgs_nwis', variable='wtd',
                            temporal_resolution='hourly', aggregation='average') == 3
    assert utils.get_var_id(conn, data_source='usgs_nwis', variable='wtd',
                            temporal_resolution='daily', aggregation='average') == 4
    assert utils.get_var_id(conn, data_source='usgs_nwis', variable='wtd',
                            temporal_resolution='instantaneous', aggregation='instantaneous') == 5
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='swe',
                            temporal_resolution='daily', aggregation='start-of-day') == 6
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='precipitation',
                            temporal_resolution='daily', aggregation='accumulated') == 7
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='precipitation',
                            temporal_resolution='daily', aggregation='total') == 8
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='precipitation', temporal_resolution='daily',
                            aggregation='total, snow-adjusted') == 9
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='temperature',
                            temporal_resolution='daily', aggregation='minimum') == 10
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='temperature',
                            temporal_resolution='daily', aggregation='maximum') == 11
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='temperature',
                            temporal_resolution='daily', aggregation='average') == 12
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='soil moisture', temporal_resolution='daily',
                            aggregation='start-of-day', depth_level=2) == 13
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='soil moisture', temporal_resolution='daily',
                            aggregation='start-of-day', depth_level=4) == 14
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='soil moisture', temporal_resolution='daily',
                            aggregation='start-of-day', depth_level=8) == 15
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='soil moisture', temporal_resolution='daily',
                            aggregation='start-of-day', depth_level=20) == 16
    assert utils.get_var_id(conn, data_source='usda_nrcs', variable='soil moisture', temporal_resolution='daily',
                            aggregation='start-of-day', depth_level=40) == 17
    cleanup_mock_hydrodata(HYDRODATA)


def test_filter_min_num_obs():
    df = pd.DataFrame({'site_id': ['101', '102', '103', '104', '105'],
                       'date1': [1, 5, 3, 4, 8], 'date2': [np.nan, 4, 2, 9, 4],
                       'date3': [np.nan, 9, 2, np.nan, 9]})

    assert len(utils.filter_min_num_obs(df, 1)) == 5
    assert len(utils.filter_min_num_obs(df, 2)) == 4
    assert len(utils.filter_min_num_obs(df, 3)) == 3


if __name__ == "__main__":
    pytest.main()


src/hf_point_data/hf_point_data.py
import pandas as pd
import sqlite3
import datetime as dt
import numpy as np
import xarray as xr

import hf_point_data.utils as utils

HYDRODATA = '/hydrodata'
DB_PATH = f'{HYDRODATA}/national_obs/point_obs.sqlite'


def get_data(data_source, variable, temporal_resolution, aggregation,
             depth_level=None,
             date_start=None, date_end=None,
             latitude_range=None, longitude_range=None,
             site_ids=None, state=None, min_num_obs=1,
             return_metadata=False, all_attributes=False):
    """
    Collect observations data into a Pandas DataFrame.

    Observations collected from HydroData for the specified data source, variable, temporal 
    resolution, and aggregation. Optional arguments can be supplied for date bounds, geography bounds,
    the minimum number of per-site observations allowed, and/or whether site metadata
    should also be returned (in a separate DataFrame).

    Parameters
    ----------
    data_source : str
        Source from which requested data originated. Currently supported: 'usgs_nwis', 'usda_nrcs', 
        'ameriflux'.   
    variable : str
        Description of type of data requested. Currently supported: 'streamflow', 'wtd', 'swe', 
        'precipitation', 'temperature', 'soil moisture', 'latent heat flux', 'sensible heat flux', 
        'shortwave radiation', 'longwave radiation', 'vapor pressure deficit', 'wind speed'.
    temporal_resolution : str
        Collection frequency of data requested. Currently supported: 'daily', 'hourly', and 'instantaneous'.
        Please see the README documentation for allowable combinations with `variable`.
    aggregation : str
        Additional information specifying the aggregation method for the variable to be returned. 
        Options include descriptors such as 'average' and 'total'. Please see the README documentation
        for allowable combinations with `variable`.
    depth_level : int
        Depth level in inches at which the measurement is taken. Necessary for `variable` = 'soil moisture'.
    date_start : str; default=None
        'YYYY-MM-DD' date indicating beginning of time range.
    date_end : str; default=None
        'YYYY-MM-DD' date indicating end of time range.
    latitude_range : tuple; default=None
        Latitude range bounds for the geographic domain; lesser value is provided first.
    longitude_range : tuple; default=None
        Longitude range bounds for the geographic domain; lesser value is provided first.
    site_ids : list; default=None
        List of desired (string) site identifiers.
    state : str; default=None
        Two-letter postal code state abbreviation.
    min_num_obs : int; default=1
        Value for the minimum number of observations desired for a site to have.
    return_metadata : bool; default=False
        Whether to additionally return a DataFrame containing site metadata.
    all_attributes : bool; default=False
        Whether to include all available attributes on returned metadata DataFrame.
    db_path : str
        Full path to location of point observations database.

    Returns
    -------
    data_df : DataFrame
        Stacked observations data for a single variable, filtered to only sites that
        (optionally) have the minimum number of observations specified, within the 
        defined geographic and/or date range.
    metadata_df : DataFrame; optional
        Metadata about the sites present in `data_df` for the desired variable.
    """

    # Create database connection
    conn = sqlite3.connect(DB_PATH)

    # Validation checks on inputs
    utils.check_inputs(data_source, variable, temporal_resolution, aggregation,
                       depth_level, return_metadata, all_attributes)

    # Get associated variable IDs for requested data types and time periods
    var_id = utils.get_var_id(conn, data_source, variable, temporal_resolution, aggregation, depth_level)

    # Get site metadata
    metadata_temp = utils.get_observations_metadata(conn, var_id,
                                                    date_start=date_start, date_end=date_end,
                                                    latitude_range=latitude_range, longitude_range=longitude_range,
                                                    site_ids=site_ids, state=state, all_attributes=all_attributes)

    if len(metadata_temp) == 0:
        raise ValueError('There are zero sites that satisfy the given parameters.')

    # Get data
    site_list = list(metadata_temp['site_id'])

    if (var_id in (1, 2, 3, 4)) | (var_id in range(6, 25)):
        data_df = utils.get_data_nc(site_list, var_id, date_start, date_end, min_num_obs)

    elif var_id == 5:
        data_df = utils.get_data_sql(conn, var_id, date_start, date_end, min_num_obs)

    # Return data
    if return_metadata == True:

        # Filter metadata down to only sites with actual data for the specific date range
        # and/or minimum observation count
        metadata_df = metadata_temp.merge(data_df['site_id'], how='right', on='site_id')
        return (data_df.reset_index().drop('index', axis=1), metadata_df.reset_index().drop('index', axis=1))

    else:
        return data_df.reset_index().drop('index', axis=1)


def get_citation_information(data_source, site_ids=None):
    """
    Print and/or return specific citation information for requested data source.

    Parameters
    ----------
    data_source : str
        Source from which data originates. Options include: 'usgs_nwis', 'usda_nrcs', and 
        'ameriflux'.
    site_ids : list; default None
        If provided, the specific list of sites to return data DOIs for. This is only
        supported if `data_source` == 'ameriflux'.

    Returns
    -------
    None or DataFrame of site-specific DOIs
        Nothing returned unless data_source == `ameriflux` and the parameter `site_list` is provided.
    """
    try:
        assert data_source in ['usgs_nwis', 'usda_nrcs', 'ameriflux']
    except:
        raise ValueError(
            f"Unexpected value of data_source, {data_source}. Supported values include 'usgs_nwis', 'usda_nrcs', and 'ameriflux'")

    if data_source == 'usgs_nwis':
        print('''Most U.S. Geological Survey (USGS) information resides in Public Domain 
              and may be used without restriction, though they do ask that proper credit be given.
              An example credit statement would be: "(Product or data name) courtesy of the U.S. Geological Survey"
              Source: https://www.usgs.gov/information-policies-and-instructions/acknowledging-or-crediting-usgs''')

    elif data_source == 'usda_nrcs':
        print('''Most information presented on the USDA Web site is considered public domain information. 
                Public domain information may be freely distributed or copied, but use of appropriate
                byline/photo/image credits is requested. 
                Attribution may be cited as follows: "U.S. Department of Agriculture"
                Source: https://www.usda.gov/policies-and-links''')

    elif data_source == 'ameriflux':
        print('''All AmeriFlux sites provided by the HydroData service follow the CC-BY-4.0 License.
                The CC-BY-4.0 license specifies that the data user is free to Share (copy and redistribute 
                the material in any medium or format) and/or Adapt (remix, transform, and build upon the 
                material) for any purpose.
            
                Users of this data must acknowledge the AmeriFlux data resource with the following statement:
                "Funding for the AmeriFlux data portal was provided by the U.S. Department of Energy Office 
                of Science."
            
                Additionally, for each AmeriFlux site used, you must provide a citation to the site's 
                data product that includes the data product DOI. The DOI for each site is included in the 
                full metadata query. Alternately, a site list can be provided to this get_citation_information
                function to return each site-specific DOI.
            
                Source: https://ameriflux.lbl.gov/data/data-policy/''')

    if site_ids is not None:
        # Create database connection
        conn = sqlite3.connect(DB_PATH)

        query = """
                SELECT site_id, doi 
                FROM sites
                WHERE site_id IN (%s)
                """ % ','.join('?'*len(site_ids))

        df = pd.read_sql_query(query, conn, params=site_ids)
        return df


src/hf_point_data/utils.py
'''
Point observations utility functions.

Note that these functions are not intended to be used stand-alone; they act as sub-processes
within the collect_observations.get_pandas_observations method.
'''
import pandas as pd
import sqlite3
import datetime as dt
import numpy as np
import xarray as xr


def check_inputs(data_source, variable, temporal_resolution, aggregation, depth_level, return_metadata, all_attributes):
    """
    Checks on inputs to get_observations function.

    Parameters
    ----------
    data_source : str
        Source from which requested data originated. Currently supported: 'usgs_nwis', 'usda_nrcs', 
        'ameriflux'.
    variable : str
        Description of type of data requested. Currently supported: 'streamflow', 'wtd', 'swe', 
        'precipitation', 'temperature', 'soil moisture', 'latent heat flux', 'sensible heat flux', 
        'shortwave radiation', 'longwave radiation', 'vapor pressure deficit', 'wind speed'.
    temporal_resolution : str
        Collection frequency of data requested. Currently supported: 'daily', 'hourly', and 'instantaneous'.
    aggregation : str
        Additional information specifying the aggregation method for the variable to be returned. 
        Options include descriptors such as 'average' and 'total'. Please see the README documentation
        for allowable combinations with `variable`.
    depth_level : int
        Depth level in inches at which the measurement is taken. Necessary for `variable` = 'soil moisture'.
    return_metadata : bool
        Whether the metadata DataFrame is also returned.
    all_attributes : bool
        If the metadata DataFrame is returned, and indication of whether the full set of site attributes
        is included or only a subset.

    Returns
    -------
    None
    """
    assert temporal_resolution in ['daily', 'hourly', 'instantaneous']
    assert variable in ['streamflow', 'wtd', 'swe', 'precipitation', 'temperature', 'soil moisture',
                        'latent heat flux', 'sensible heat flux', 'shortwave radiation', 'longwave radiation',
                        'vapor pressure deficit', 'wind speed']
    assert aggregation in ['average', 'instantaneous', 'total', 'total, snow-adjusted',
                           'start-of-day', 'accumulated', 'minimum', 'maximum']
    assert data_source in ['usgs_nwis', 'usda_nrcs', 'ameriflux']
    assert depth_level in [2, 4, 8, 20, 40, None]

    if variable == 'soil moisture':
        assert depth_level is not None

    if return_metadata == False:
        assert all_attributes == False


def get_var_id(conn, data_source, variable, temporal_resolution, aggregation, depth_level=None):
    """
    Return mapped var_id.

    Parameters
    ----------
    conn : Connection object
        The Connection object associated with the SQLite database to 
        query from. 
    data_source : str
        Source from which requested data originated. Currently supported: 'usgs_nwis', 'usda_nrcs', 
        'ameriflux'.    
    variable : str
        Description of type of data requested. Currently supported: 'streamflow', 'wtd', 'swe', 
        'precipitation', 'temperature', 'soil moisture', 'latent heat flux', 'sensible heat flux', 
        'shortwave radiation', 'longwave radiation', 'vapor pressure deficit', 'wind speed'.
    temporal_resolution : str
        Collection frequency of data requested. Currently supported: 'daily', 'hourly', and 'instantaneous'.
    aggregation : str
        Additional information specifying the aggregation method for the variable to be returned. 
        Options include descriptors such as 'average' and 'total'. Please see the README documentation
        for allowable combinations with `variable`.
    depth_level : int
        Depth level in inches at which the measurement is taken. Necessary for variable='soil moisture'.

    Returns
    -------
    var_id : int
        Integer variable ID associated with combination of `data_source`, `variable`, `temporal_resolution`,
        and `aggregation`.
    """
    if variable == 'soil moisture':
        query = """
                SELECT var_id 
                FROM variables
                WHERE data_source = ?
                    AND variable = ?
                    AND temporal_resolution = ?
                    AND aggregation = ?
                    AND depth_level = ?
                """
        param_list = [data_source, variable, temporal_resolution, aggregation, depth_level]

    else:
        query = """
                SELECT var_id 
                FROM variables
                WHERE data_source = ?
                    AND variable = ?
                    AND temporal_resolution = ?
                    AND aggregation = ?
                """
        param_list = [data_source, variable, temporal_resolution, aggregation]

    try:
        result = pd.read_sql_query(query, conn, params=param_list)
        return int(result['var_id'][0])
    except:
        raise ValueError(
            'The provided combination of data_source, variable, temporal_resolution, and aggregation is not currently supported.')


def get_dirpath(var_id):
    """
    Map variable with location of data on /hydrodata.

    Parameters
    ----------
    var_id : int
        Integer variable ID associated with combination of `data_source`, 
        `variable`, `temporal_resolution`, and `aggregation`.

    Returns
    -------
    dirpath : str
        Directory path for observation data location.
    """
    dirpath_map = {1: '/hydrodata/national_obs/streamflow/data/hourly',
                   2: '/hydrodata/national_obs/streamflow/data/daily',
                   3: '/hydrodata/national_obs/groundwater/data/hourly',
                   4: '/hydrodata/national_obs/groundwater/data/daily',
                   5: '',
                   6: '/hydrodata/national_obs/swe/data/daily',
                   7: '/hydrodata/national_obs/point_meteorology/NRCS_precipitation/data/daily',
                   8: '/hydrodata/national_obs/point_meteorology/NRCS_precipitation/data/daily',
                   9: '/hydrodata/national_obs/point_meteorology/NRCS_precipitation/data/daily',
                   10: '/hydrodata/national_obs/point_meteorology/NRCS_temperature/data/daily',
                   11: '/hydrodata/national_obs/point_meteorology/NRCS_temperature/data/daily',
                   12: '/hydrodata/national_obs/point_meteorology/NRCS_temperature/data/daily',
                   13: '/hydrodata/national_obs/soil_moisture/data/daily',
                   14: '/hydrodata/national_obs/soil_moisture/data/daily',
                   15: '/hydrodata/national_obs/soil_moisture/data/daily',
                   16: '/hydrodata/national_obs/soil_moisture/data/daily',
                   17: '/hydrodata/national_obs/soil_moisture/data/daily',
                   18: '/hydrodata/national_obs/ameriflux/data/hourly',
                   19: '/hydrodata/national_obs/ameriflux/data/hourly',
                   20: '/hydrodata/national_obs/ameriflux/data/hourly',
                   21: '/hydrodata/national_obs/ameriflux/data/hourly',
                   22: '/hydrodata/national_obs/ameriflux/data/hourly',
                   23: '/hydrodata/national_obs/ameriflux/data/hourly',
                   24: '/hydrodata/national_obs/ameriflux/data/hourly'}

    return dirpath_map[var_id]


def get_observations_metadata(conn, var_id, date_start=None, date_end=None,
                              latitude_range=None, longitude_range=None,
                              site_ids=None, state=None, all_attributes=False):
    """
    Build DataFrame with site attribute metadata information.

    Parameters
    ----------
    conn : Connection object
        The Connection object associated with the SQLite database to 
        query from. 
    var_id : int
        Integer variable ID associated with combination of `data_source`, `variable`, `temporal_resolution`,
        and `aggregation`.
    date_start : str; default=None
        'YYYY-MM-DD' format date indicating beginning of time range.
    date_end : str; default=None
        'YYYY-MM-DD' format date indicating end of time range.
    latitude_range : tuple; default=None
        Latitude range bounds for the geographic domain; lesser value is provided first.
    longitude_range : tuple; default=None
        Longitude range bounds for the geographic domain; lesser value is provided first.
    site_ids : list; default=None
        List of desired (string) site identifiers.
    state : str; default=None
        Two-letter postal code state abbreviation.
    all_attributes : bool; default=False
        Whether to include all available attributes on returned DataFrame.

    Returns
    -------
    DataFrame
        Site-level DataFrame of attribute metadata information.

    Notes
    -----
    The returned field 'record_count' is OVERALL record count. Filtering of metadata 
    only applies at the site level, so only sites within the provided bounds 
    (space and time) are included. The record count does not reflect any filtering 
    at the data/observation level.
    """
    if var_id in (1, 2):
        attribute_table = 'streamgauge_attributes'
    elif var_id in (3, 4, 5):
        attribute_table = 'well_attributes'
    elif var_id in range(6, 18):
        attribute_table = 'snotel_station_attributes'
    elif var_id in range(18, 25):
        attribute_table = 'flux_tower_attributes'

    param_list = [var_id]

    # Date start
    if date_start != None:
        date_start_query = """ AND last_date_data_available >= ?"""
        param_list.append(date_start)
    else:
        date_start_query = """"""

    # Date end
    if date_end != None:
        date_end_query = """ AND first_date_data_available <= ?"""
        param_list.append(date_end)
    else:
        date_end_query = """"""

    # Latitude
    if latitude_range != None:
        lat_query = """ AND latitude BETWEEN ? AND ?"""
        param_list.append(latitude_range[0])
        param_list.append(latitude_range[1])
    else:
        lat_query = """"""

    # Longitude
    if longitude_range != None:
        lon_query = """ AND longitude BETWEEN ? AND ?"""
        param_list.append(longitude_range[0])
        param_list.append(longitude_range[1])
    else:
        lon_query = """"""

    # Site ID
    if site_ids != None:
        site_query = """ AND s.site_id IN (%s)""" % ','.join('?'*len(site_ids))
        for s in site_ids:
            param_list.append(s)
    else:
        site_query = """"""

    # State
    if state != None:
        state_query = """ AND state == ?"""
        param_list.append(state)
    else:
        state_query = """"""

    query_full = """
                SELECT * 
                FROM
                (SELECT *
                    FROM sites s
                    INNER JOIN observations o
                    ON s.site_id = o.site_id AND o.var_id == ?
                    WHERE first_date_data_available <> 'None' 
                """ + date_start_query + date_end_query + lat_query + lon_query + site_query + state_query + """ GROUP BY s.site_id)
                INNER JOIN {}
                USING (site_id)
                """.format(attribute_table)

    query_subset = """
                    SELECT s.site_id, s.site_name, s.site_type, s.agency, s.state, 
                           s.latitude, s.longitude, o.var_id, o.first_date_data_available,
                           o.last_date_data_available, o.record_count, o.file_path
                    FROM sites s
                    INNER JOIN observations o
                    ON s.site_id = o.site_id AND o.var_id == ?
                    WHERE first_date_data_available <> 'None' 
                    """ + date_start_query + date_end_query + lat_query + lon_query + site_query + state_query

    if all_attributes == True:
        query = query_full
    else:
        query = query_subset

    df = pd.read_sql_query(query, conn, params=param_list)

    return df


def convert_to_pandas(ds):
    """
    Convert xarray DataSet to pandas DataFrame.

    Parameters
    ----------
    ds : DataSet
        xarray DataSet containing stacked observations data for a 
        single variable. 
    var_id : int
        Integer variable ID associated with combination of `variable`, `temporal_resolution`,
        and `aggregation`.

    Returns
    -------
    DataFrame
        Stacked observations data for a single variable.
    """
    sites = pd.Series(ds['site'].to_numpy())
    dates = pd.Series(ds['date'].to_numpy()).astype(str)
    data = ds.to_numpy()

    df = pd.DataFrame(data, columns=dates)
    df['site_id'] = sites

    # Reorder columns to put site_id first
    cols = df.columns.tolist()
    cols = cols[-1:] + cols[:-1]
    df = df[cols]

    return df


def filter_min_num_obs(df, min_num_obs):
    """
    Filter to only sites which have a minimum number of observations.

    This filtering is done after the observations are subset by time, so these
    observation counts will only filter out sites if the number of observations *within 
    that time range* is not satisfied.

    Parameters
    ----------
    df : DataFrame
        Stacked observations data for a single variable.
    min_num_obs : int
        Value for the minimum number of observations desired for a site to have.

    Returns
    -------
    DataFrame
        Stacked observations data for a single variable, filtered to only sites that
        have the minimum number of observations specified.
    """
    dfc = df.copy()

    # count number of non-na observations
    dfc['num_obs'] = dfc.count(axis=1) - 1

    # filter based on this field
    df_filtered = dfc[dfc['num_obs'] >= min_num_obs]

    return df_filtered


def get_data_nc(site_list, var_id, date_start, date_end, min_num_obs):
    """
    Get observations data for data that is stored in NetCDF files.

    Parameters
    ----------
    site_list : list
        List of site IDs to query observations data for.
    var_id : int
        Integer variable ID associated with combination of `data_source`,
        `variable`, `temporal_resolution`, and `aggregation`.
    date_start : str; default=None
        'YYYY-MM-DD' format date indicating beginning of time range.
    date_end : str; default=None
        'YYYY-MM-DD' format date indicating end of time range.
    min_num_obs : int
        Value for the minimum number of observations desired for a site to have.

    Returns
    -------
    DataFrame
        Stacked observations data for a single variable, filtered to only sites that
        have the minimum number of observations specified.
    """
    dirpath = get_dirpath(var_id)
    file_list = [f'{dirpath}/{site}.nc' for site in site_list]

    varname_map = {'1': 'streamflow', '2': 'streamflow', '3': 'wtd', '4': 'wtd', '5': 'wtd',
                   '6': 'swe', '7': 'precip_acc', '8': 'precip_inc', '9': 'precip_inc_sa',
                   '10': 'temp_min', '11': 'temp_max', '12': 'temp_avg',
                   '13': 'sms_2in', '14': 'sms_4in', '15': 'sms_8in', '16': 'sms_20in', '17': 'sms_40in',
                   '18': 'latent heat flux', '19': 'sensible heat flux', '20': 'shortwave radiation',
                   '21': 'longwave radiation', '22': 'vapor pressure deficit', '23': 'air temperature',
                   '24': 'wind speed'}

    varname = varname_map[str(var_id)]

    if date_start != None:
        date_start_dt = np.datetime64(date_start)
    if date_end != None:
        date_end_dt = np.datetime64(date_end)

    print('collecting data...')

    for i in range(len(site_list)):

        # open single site file
        temp = xr.open_dataset(file_list[i])[varname]

        # make date variable name consistent
        date_var = list(temp.coords)[0]
        temp = temp.rename({date_var: 'datetime'})

        # convert date string to datetime values
        temp['datetime'] = pd.DatetimeIndex(temp['datetime'].values)

        # subset to only observations within desired time range
        if (date_start == None) and (date_end == None):
            temp_wy = temp
        elif (date_start == None) and (date_end != None):
            temp_wy = temp.sel(datetime=(temp.datetime <= date_end_dt))
        elif (date_start != None) and (date_end == None):
            temp_wy = temp.sel(datetime=(temp.datetime >= date_start_dt))
        elif (date_start != None) and (date_end != None):
            temp_wy = temp.sel(datetime=(temp.datetime >= date_start_dt) & (temp.datetime <= date_end_dt))

        if i == 0:
            ds = temp_wy
        else:
            ds = xr.concat([ds, temp_wy], dim='site')

    if len(site_list) == 1:
        ds = ds.expand_dims(dim='site')

    ds = ds.assign_coords({'site': (site_list)})
    ds = ds.rename({'datetime': 'date'})

    print('data collected.')

    data_df = convert_to_pandas(ds)
    final_data_df = filter_min_num_obs(data_df, min_num_obs)

    return final_data_df


def get_data_sql(conn, var_id, date_start, date_end, min_num_obs):
    """
    Get observations data for data that is stored in a SQL table.

    Parameters
    ----------
    conn : Connection object
        The Connection object associated with the SQLite database to 
        query from. 
    var_id : int
        Integer variable ID associated with combination of `data_source`, 
        `variable`, `temporal_resolution`, and `aggregation`.
    date_start : str; default=None
        'YYYY-MM-DD' format date indicating beginning of time range.
    date_end : str; default=None
        'YYYY-MM-DD' format date indicating end of time range.
    min_num_obs : int
        Value for the minimum number of observations desired for a site to have.

    Returns
    -------
    DataFrame
        Stacked observations data for a single variable, filtered to only sites that
        have the minimum number of observations specified.
    """
    assert var_id == 5

    # Note:
    #   pumping_status == '1' --> Static (not pumping)
    #   pumping_status == 'P' --> Pumping
    #   pumping_status == '' --> unknown (not reported)

    if (date_start == None) and (date_end == None):
        date_query = """"""
        param_list = [min_num_obs]
    elif (date_start == None) and (date_end != None):
        date_query = """ WHERE w.date <= ?"""
        param_list = [date_end, min_num_obs, date_end]
    elif (date_start != None) and (date_end == None):
        date_query = """ WHERE w.date >= ?"""
        param_list = [date_start, min_num_obs, date_start]
    elif (date_start != None) and (date_end != None):
        date_query = """ WHERE w.date >= ? AND w.date <= ?"""
        param_list = [date_start, date_end, min_num_obs, date_start, date_end]

    query = """
            SELECT w.site_id, w.date, w.wtd, w.pumping_status, num_obs
            FROM wtd_discrete_data AS w
            INNER JOIN (SELECT w.site_id, COUNT(*) AS num_obs
                FROM wtd_discrete_data AS w
                """ + date_query + """
                GROUP BY site_id
                HAVING num_obs >= ?) AS c
            ON w.site_id = c.site_id
            """ + date_query

    df = pd.read_sql_query(query, conn, params=param_list)

    return df
